2016/presentations/Python_for_Networking_Devices/text.md

```diff
@@ -0,0 +1,139 @@
+the device - the change is not transactional.
+
+Nowadays Netconf is the new hype in the networking world. With underlying SSH transport it supports the same security
+feature set, but in addition it offers support  for transactions, structured data and error returns. Netconf allows to
+submit multiple changes at a time via an RPC call, encoded as XML and returns an XML object with the result in return.
+
+Even though Netconf is an RFC standard and supposedly vendors should implement it in a similar fashion - they don't.
+Not every vendor supports the full Netconf feature set, if they offer support for it at all, in which case you are
+stuck with managing your devices line by line via the CLI after all.
+
+In addition to configuring specific functionality on the device, 'show commands' are used to retrieve operational or
+state data of the router. For example, output like this can be retrieved upon requesting the current ARP table on an
+Arista device:
+```
+Address         Age (min)  Hardware Addr   Interface
+10.220.88.1             0  001f.9e92.16fb  Vlan1, Ethernet1
+10.220.88.21            0  1c6a.7aaf.576c  Vlan1, not learned
+10.220.88.28            0  5254.00ee.446c  Vlan1, not learned
+10.220.88.29            0  5254.0098.69b6  Vlan1, not learned
+10.220.88.30            0  5254.0092.13bb  Vlan1, not learned
+10.220.88.38            0  0001.00ff.0001  Vlan1, not learned
+```
+
+On a Juniper device, it will look more like this:
+```
+MAC Address       Address         Name                      Interface           Flags
+00:1f:9e:92:16:fb 10.220.88.1     10.220.88.1               vlan.0              none
+00:19:e8:45:ce:80 10.220.88.22    10.220.88.22              vlan.0              none
+f0:ad:4e:01:d9:33 10.220.88.100   10.220.88.100             vlan.0              none
+Total entries: 3
+```
+
+Not only can the show commands to execute on the device differ per vendor, they also provide different text format
+output, which in case of CLI access has to be parsed individually. Netconf like access methods support the retrieval
+of structured data, which is slightly better, but this still doesn’t cover fields not provided in one vendor vs.
+another (like for example the lack of an age timer in the Juniper output above).
+
+##Generic Access Libraries
+
+To access devices directly via their SSH or Netconf interface, generic python libraries such as
+[pexpect](https://github.com/pexpect/pexpect), [paramiko](https://github.com/paramiko/paramiko) and
+[ncclient](https://github.com/ncclient/ncclient) can be used. They allow for programmatic access from scripts
+(for network engineers who are trying to figure out how to code ;) ) but don't provide any ease of use in terms of
+vendor specificity. You will still have to deal with the little differences the vendors bring, different login
+procedured or additional escape chars). You will take care of different lines of configuration or show commands you
```

Comment:
I think it should be `different login procedures` instead of `different login procedured`